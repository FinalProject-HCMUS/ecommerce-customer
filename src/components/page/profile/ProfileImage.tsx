import { useRef } from "react"
import { motion } from "framer-motion"
import { Camera } from "lucide-react"
import type React from "react"

interface ProfileImageProps {
  previewImage: string
  isEditing: boolean
  onImageChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}

const ProfileImage : React.FC<ProfileImageProps> =  ({ previewImage, isEditing, onImageChange }) => {
  const fileInputRef = useRef<HTMLInputElement>(null)

  const handleImageClick = (): void => {
    if (isEditing && fileInputRef.current) {
      fileInputRef.current.click()
    }
  }

  return (
    <div className="relative group cursor-pointer" onClick={handleImageClick}>
      <motion.div
        whileHover={{ scale: 1.05 }}
        whileTap={{ scale: 0.95 }}
        className="h-32 w-32 sm:h-40 sm:w-40 rounded-full border-4 border-white overflow-hidden bg-white shadow-lg"
      >
        <img src={previewImage || "/placeholder.svg"} alt="Profile" className="h-full w-full object-cover" />
        {isEditing && (
          <div className="absolute inset-0 bg-black bg-opacity-50 flex items-center justify-center opacity-0 group-hover:opacity-100 transition-opacity">
            <Camera className="h-8 w-8 text-white" />
          </div>
        )}
      </motion.div>
      <input
        type="file"
        ref={fileInputRef}
        onChange={onImageChange}
        accept="image/*"
        className="hidden"
        disabled={!isEditing}
      />
    </div>
  )
}

export default ProfileImage
