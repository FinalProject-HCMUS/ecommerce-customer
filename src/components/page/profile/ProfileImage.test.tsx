import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ProfileImage from './ProfileImage';

// Mock framer-motion to avoid animation-related issues in tests
jest.mock('framer-motion', () => ({
  motion: {
    div: ({
      children,
      className,
      ...props
    }: {
      children?: React.ReactNode;
      className?: string;
      [key: string]: any;
    }) => (
      <div className={className} data-testid="motion-div" {...props}>
        {children}
      </div>
    ),
  },
}));

describe('ProfileImage Component', () => {
  const mockOnImageChange = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders with default props', () => {
    render(
      <ProfileImage
        previewImage=""
        isEditing={false}
        onImageChange={mockOnImageChange}
      />
    );

    // Check if the component renders
    expect(screen.getByAltText('Profile')).toBeInTheDocument();

    // Should use placeholder image when previewImage is empty
    expect(screen.getByAltText('Profile')).toHaveAttribute(
      'src',
      '/placeholder.svg'
    );

    // Camera icon should not be visible when not editing
    expect(
      screen.queryByRole('img', { name: /camera/i })
    ).not.toBeInTheDocument();
  });

  test('renders with provided image', () => {
    render(
      <ProfileImage
        previewImage="test-image.jpg"
        isEditing={false}
        onImageChange={mockOnImageChange}
      />
    );

    // Should use the provided image
    expect(screen.getByAltText('Profile')).toHaveAttribute(
      'src',
      'test-image.jpg'
    );
  });

  test('clicking does not trigger file input when not in editing mode', () => {
    render(
      <ProfileImage
        previewImage="test-image.jpg"
        isEditing={false}
        onImageChange={mockOnImageChange}
      />
    );

    // Mock the click method of the hidden file input
    const clickMock = jest.fn();
    HTMLInputElement.prototype.click = clickMock;

    // Click on the image container
    fireEvent.click(screen.getByTestId('motion-div'));

    // File input click should not be triggered
    expect(clickMock).not.toHaveBeenCalled();
  });

  test('clicking does not trigger file input when uploading', () => {
    render(
      <ProfileImage
        previewImage="test-image.jpg"
        isEditing={true}
        isUploading={true}
        onImageChange={mockOnImageChange}
      />
    );

    // Mock the click method of the hidden file input
    const clickMock = jest.fn();
    HTMLInputElement.prototype.click = clickMock;

    // Click on the image container
    fireEvent.click(screen.getByTestId('motion-div'));

    // File input click should not be triggered
    expect(clickMock).not.toHaveBeenCalled();
  });

  test('matches snapshot', () => {
    const { container } = render(
      <ProfileImage
        previewImage="test-image.jpg"
        isEditing={true}
        onImageChange={mockOnImageChange}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
