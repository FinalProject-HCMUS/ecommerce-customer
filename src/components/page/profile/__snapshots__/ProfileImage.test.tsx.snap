// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileImage Component matches snapshot 1`] = `
<div>
  <div
    class="relative group cursor-pointer"
  >
    <div
      class="h-32 w-32 sm:h-40 sm:w-40 rounded-full border-4 border-white overflow-hidden bg-white shadow-lg"
      data-testid="motion-div"
      whilehover="[object Object]"
      whiletap="[object Object]"
    >
      <img
        alt="Profile"
        class="h-full w-full object-cover"
        src="test-image.jpg"
      />
      <div
        class="absolute inset-0 bg-black bg-opacity-50 flex items-center justify-center opacity-0 group-hover:opacity-100 transition-opacity"
      >
        <svg
          class="lucide lucide-camera h-8 w-8 text-white"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.5 4h-5L7 7H4a2 2 0 0 0-2 2v9a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V9a2 2 0 0 0-2-2h-3l-2.5-3z"
          />
          <circle
            cx="12"
            cy="13"
            r="3"
          />
        </svg>
      </div>
    </div>
    <input
      accept="image/*"
      class="hidden"
      type="file"
    />
  </div>
</div>
`;
